crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//! only the `as <type>` parts are new:
+//!
+//! ```ignore
+//!         .map_index(|(c, (date, avg))| {
+//!             (
+//!                 (c.clone(), (date / 12) as i32, (date % 12 + 1) as u8),
+//!                 avg.unwrap(),
+//!             )
+//!         });
+//! ```
+//!
+//! Then we can use [`join_index`](`Stream::join_index`) to do the join and
+//! [`inspect`](`Stream::inspect`) to print the results, as shown below.  The
+//! only subtlety is the use of `Some(...)`, which is here because the closure
+//! passed to `join_index` must return an iterable collection and `Some(x)` is a
+//! convenient way to construct a collection with exactly one value in it.
+//!
+//! ```ignore
+//!    monthly_totals
+//!         .join_index(&moving_averages, |(c, y, m), cur, avg| {
+//!             Some(((c.clone(), *y, *m), (*cur, *avg)))
+//!         })
+//!         .inspect(|data| {
+//!             data.iter().for_each(|((c, y, m), (cur, avg), w)| {
+//!                 println!("{c:16} {y}-{m:02} {cur:10} {avg:10}: {w:+}")
+//!             })
+//!         });
+//!  ```
+//!
+//! The whole program is in `tutorial7.rs`.  If we run it, it prints both per-month
+//! vaccination numbers and 3-month moving averages:
+//!  
+//!  ```text
+//! England          2021-01    5600174    5600174: +1
+//! England          2021-02    9377418    7488796: +1
+//! England          2021-03   11861175    8946255: +1
+//! England          2021-04   11288945   10842512: +1
+//! England          2021-05   13772946   12307688: +1
+//! England          2021-06   10944915   12002268: +1
+//! ...
+//! ```
+//!
+//! ### Finding months with the most vaccinations
+//!
+//! Suppose we want to find the months when the most vaccinations occurred in
+//! each country.  [`Stream`] has a ready-made method [`Stream::topk_desc`] for
+//! this, which we simply pass the number of top records to keep per group.  (If
+//! we only want the greatest value, rather than the top-`k` for `k > 1`, then
+//! [`Stream::aggregate`] with the [`Max`] aggregator also works.)
+//!
+//! There is one tricky part.  To use [`topk_desc`](`Stream::topk_desc`), we
+//! must re-index our data so that the country is the key (used for grouping)
+//! and the number of vaccinations is the value. But, if we do that in the most
+//! obvious way, we end up with just the number of vaccinations as the result,
+//! whereas we probably want to know the year and month that that number of
+//! occurred as well.
+//!
+//! One way to recover the year and month is to join against the original data.
+//! We can do it without another join by defining a type that is ordered by
+//! vaccination count but also contains the year and month.  Taking advantage of
+//! how Rust derives [`Ord`] lexicographically, that's as simple as:
+//!
+//! ```rust
+//! # use size_of::SizeOf;
+//!
+//! #[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, SizeOf)]
```

Comment:
some of these 'derive's are obvious, but some aren't.
I think that in an advanced tutorial they deserve a mention.